packages/backend/server/src/core/version/service.ts

```diff
@@ -0,0 +1,68 @@
+import assert from 'node:assert';
+
+import { Injectable, Logger } from '@nestjs/common';
+import semver from 'semver';
+
+import { Runtime, UnsupportedClientVersion } from '../../base';
+
+@Injectable()
+export class VersionService {
+  private readonly logger = new Logger(VersionService.name);
+
+  constructor(private readonly runtime: Runtime) {}
+
+  private async getRecommendedVersion(versionRange: string) {
+    try {
+      const range = new semver.Range(versionRange);
+      const versions = range.set
+        .flat()
+        .map(c => c.semver)
+        .toSorted((a, b) => semver.rcompare(a, b));
+      return versions[0]?.toString();
+    } catch {
+      return semver.valid(semver.coerce(versionRange));
+    }
+  }
+
+  async checkVersion(clientVersion?: any) {
```

Comment:
we might need the channel control, cross channel connections should be forbidden